### PR TITLE
feat(SearchResults): backport instantsearch.js getRefinement method

### LINKED ISSUE #195: Ability to easily list refined attributes
To conditionally display the list of current applied filters OR/AND a `clearRefinements` button.

### PR DESCRIPTION
**Pull request by bobylito**
Fixes #195

### CODE REVIEW

**Comment by bobylito** [comment]
cc @redox @Morhaus

File: src/SearchResults/index.js [hunk 1]
```diff
@@ -46,9 +47,19 @@ var generateHierarchicalTree = require('./generate-hierarchical-tree');
 /**
  * @typedef SearchResults.FacetValue
  * @type {object}
- * @property {value} string the facet value itself
- * @property {count} number times this facet appears in the results
- * @property {isRefined} boolean is the facet currently selected
+ * @property {string } value the facet value itself
```
Comment (Morhaus) [inline src/SearchResults/index.js]: I don't know if it changes anything, but there's an extra space after `string`.

File: src/SearchResults/index.js [hunk 2]
```diff
@@ -46,9 +47,19 @@ var generateHierarchicalTree = require('./generate-hierarchical-tree');
...
+ * @property {number} count times this facet appears in the results
+ * @property {boolean} isRefined is the facet currently selected
+ */
+
+/**
+ * @typedef Refinement
+ * @typedef {object}
```
Comment (Morhaus) [inline src/SearchResults/index.js]: You mean @type ?

File: src/SearchResults/index.js [hunk 3]
```diff
@@ -620,4 +631,91 @@ function getFacetStatsIfAvailable(facetList, facetName) {
...
+  return {
+    type: type,
+    attributeName: attributeName,
+    name: name,
+    count: count || 0,
+    exhaustive: exhaustive || false
+  };
+}
+
+function getHierarchicalRefinement(state, type, attributeName, name, resultsFacets) {
```
Comment (Morhaus) [inline src/SearchResults/index.js]: ?

File: src/SearchResults/index.js [hunk 4]
```diff
@@ -620,4 +631,91 @@ function getFacetStatsIfAvailable(facetList, facetName) {
...
+
+  forEach(state.tagRefinements, function(name) {
+    res.push({type: 'tag', attributeName: '_tags', name});
+  });
+
+  return res;
+};
+
+function getRefinement(state, type, attributeName, name, resultsFacets) {
+  var name = name;
```
Comment (Morhaus) [inline src/SearchResults/index.js]: Is this in order to avoid mutating arguments?

File: src/algoliasearch.helper.js [hunk 5]
```diff
@@ -145,7 +145,7 @@ AlgoliaSearchHelper.prototype.search = function() {
  * }
  */
 AlgoliaSearchHelper.prototype.searchOnce = function(options, cb) {
-  var tempState = this.state.setQueryParameters(options);
+  var tempState = this.state.setQueryParameters(options || {});
```
Comment (Morhaus) [inline src/algoliasearch.helper.js]: If options is not set, then maybe we should just do `options ? this.state.setQueryParameters(options) : this.state`?

File: test/replayTools.js [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+      this,
+      Array.prototype.slice.call(arguments)
+    ).then(function(contentAndState) {
+      savedParameters.content = contentAndState._originalResponse;
+      savedParameters.state = contentAndState.state;
+      savedParameters.error = null;
+      return contentAndState;
+    }, function(error) {
+      savedParameters.content = null;
+      savedParameters.state = null; // FIXME can't access the actual state without reimplementation
```
Comment (Morhaus) [inline test/replayTools.js]: You could maybe compute the state beforehand then pass it as the first parameter to searchOnce ?

File: src/SearchResults/index.js (hunk 1 above)
Comment (bobylito) [inline src/SearchResults/index.js]: Will fix as it breaks the highlighting of github 👍 But it doesn't create issues when rendering the doc.

File: src/SearchResults/index.js (hunk 2 above)
Comment (bobylito) [inline src/SearchResults/index.js]: 👍

File: src/SearchResults/index.js (hunk 4 above)
Comment (bobylito) [inline src/SearchResults/index.js]: I will refactor that :/

File: src/SearchResults/index.js (hunk 3 above)
Comment (bobylito) [inline src/SearchResults/index.js]: Forgotten refactoring oups 👍

File: src/algoliasearch.helper.js (hunk 5 above)
Comment (bobylito) [inline src/algoliasearch.helper.js]: 👍

**Pull request merged**